pkg/kv/kvserver/replica_rangefeed.go

```diff
@@ -341,7 +334,7 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	r.rangefeedMu.Lock()
 	p := r.rangefeedMu.proc
 	if p != nil {
-		reg, filter := p.Register(span, startTS, catchUpIter, withDiff, stream, errC)
+		reg, filter := p.Register(span, startTS, catchUpIter, withDiff, stream, func() { r.maybeDisconnectEmptyRangefeed(p) }, done)
```

Comment:
Why aren't we doing `maybeDisconnectEmptyRangefeed()` by attaching a `WhenReady()` to the future? That would avoid having to plumb down yet another callback.

Reply:
Excellent question.  I almost undid all these changes in favor of WhenReady... Alas, it's not that simple.
There is some annoyance around setting WhenReady on a future that's already ready; this executes fn inline, thus the caller must be careful around mutexes.  Naively, adding done.WhenReady above will deadlock because r.maybeDisconnectEmptyRangefeed will acquire r.rangefeedMu, which is already head.
Annoyances aside, the *real* issue is that this whole registration/unregistration business is very much async.
A signal is sent to the processor to unregister registration.  Crucially, r.maybeDisconnect only disconnects if the number of registration in the processor is 0.  But to get the number of registration, you (async) request processor to return how many registration it has.  All of this is to say: you can't disconnect unless you already un-registred.    However, this future completes before any of the above can take place.  Thus, when maybeDisconnect runs, the count would still be 1.  I've been going back and forth on this, but this is the least invasive way I can come up.

Reply:
Makes sense.